packages/lib/src/load_plugin_list.ts

```diff
@@ -0,0 +1,184 @@
+ * @returns Promise that resolves when plugin discovery is complete
+ */
+async function findLocalPlugins(pluginList: Map<string, string>, pluginListPath: string) {
+	// Check folders for plugins
+	const pluginFolders = [
+		"plugins",
+		"external_plugins",
+	];
+
+	let has_changed = 0;
+	const ignoredFolders = new Set(["node_modules", ".git", "dist", "build", "disabled"]);
+	const maxDepth = 3;
+
+	for (const pluginFolder of pluginFolders) {
+		let pluginNames;
+		try {
+			pluginNames = await fs.readdir(pluginFolder);
+		} catch (err: any) {
+			if (err.code === "ENOENT") {
+				// Skip if folder doesn't exist
+				continue;
+			}
+			throw err;
+		}
+		for (const pluginName of pluginNames) {
+			const pluginPath = path.join(pluginFolder, pluginName);
+			if (pluginList.has(pluginName)) {
+				// If the one in the list is not the same as this one, warn
+				if (pluginList.get(pluginName) !== path.resolve(pluginPath)) {
+					logger.warn(
+						`${pluginName} is already in the plugin list, but with a different path - ` +
+						`using ${pluginList.get(pluginName)}`
+					);
+				}
+				continue;
+			}
+			const stats = await fs.stat(pluginPath);
+			if (stats.isDirectory()) {
+				// Check for package.json and plugin keyword
+				if (await checkPackageJson(pluginPath)) {
+					pluginList.set(pluginName, path.resolve(pluginPath));
+					logger.info(`Added ${pluginName} from ${pluginFolder}`);
+					has_changed += 1;
+					continue;
+				}
+
+				if (pluginPath.split(path.sep).length < maxDepth && !ignoredFolders.has(pluginName)) {
+					logger.warn(`Possible monorepo detected, scanning ${pluginPath} for clusterio-plugins`);
+					pluginFolders.push(pluginPath);
+				}
+			}
+		}
+	}
+
+	if (has_changed > 0) {
+		await libFileOps.safeOutputFile(pluginListPath, JSON.stringify([...pluginList], null, "\t"));
+	}
+}
+
+function getPluginName(requireSpec: string) {
+	const context = vm.createContext({ require: require, pluginInfo: null });
+	const code = `pluginInfo = require(${JSON.stringify(requireSpec)}).plugin;`;
+	vm.runInContext(code, context);
+	return context.pluginInfo.name;
+}
+
+/**
+ * Find NPM packages in the root package.json that satisfies the requirements for plugins.
+ *
+ * @param pluginList - Map of plugin names to their file paths
+ * @param pluginListPath - Path to the JSON file storing the plugin list
+ * @returns Promise that resolves when plugin discovery is complete
+ */
+async function findNpmPlugins(pluginList: Map<string, string>, pluginListPath: string) {
+	let dependencies;
+	try {
+		const rootPackageJson = JSON.parse(await fs.readFile(path.resolve("package.json"), { encoding: "utf8" }));
+		dependencies = rootPackageJson.dependencies;
+	} catch (err: any) {
+		if (err.code === "ENOENT") {
+			// Skip if package.json doesn't exist
+			return;
+		}
+		throw err;
+	}
+
+	let changed = 0;
+	for (const [packageName, packageVersion] of Object.entries(dependencies)) {
```

Comment:
`dependencies` is undefined in a dev install of clusterio where only `devDependencies` exists. Need an early return for undefined or null dependencies. (`if (!dependencies) return;`)